pkg/handler/collector/github/github.go

```diff
@@ -0,0 +1,206 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/github"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/logging"
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+)
+
+const (
+	CollectorGitHub = "GitHubCollector"
+)
+
+type githubCollector struct {
+	poll     bool
+	interval time.Duration
+	token    string
+	client   *github.Client
+	owner    string
+	repo     string
+	tag      string
+	tagMap   map[string]string
+}
+
+type GithubCollectorOpts struct {
+	poll     bool
+	interval time.Duration
+	token    string
+	client   *github.Client
+	owner    string
+	repo     string
+	tag      string
+}
+
+func NewGitHubCollector(ctx context.Context, gco GithubCollectorOpts) (*githubCollector, error) {
+	if err := validateOpts(gco); err != nil {
+		return nil, err
+	}
+	// Authenticate with GitHub
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: gco.token},
+	)
+	if gco.client == nil {
+		tc := oauth2.NewClient(ctx, ts)
+		gco.client = github.NewClient(tc)
+	}
+
+	return &githubCollector{
+		poll:     gco.poll,
+		interval: gco.interval,
+		token:    gco.token,
+		client:   gco.client,
+		owner:    gco.owner,
+		repo:     gco.repo,
+		tag:      gco.tag,
+		tagMap:   map[string]string{},
+	}, nil
+}
+
+func validateOpts(gco GithubCollectorOpts) error {
+	if gco.owner == "" {
+		return fmt.Errorf("expected to receive GitHub owner")
+	}
+	if gco.repo == "" {
+		return fmt.Errorf("expected to receive GitHub repo")
+	}
+	if gco.token == "" {
+		return fmt.Errorf("expected to receive GitHub token")
+	}
+	return nil
+}
+
+// RetrieveArtifacts collects the metadata documents and assets from a specified Github repository's release
+func (g *githubCollector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
```

Comment:
My suggestion.


1. Use `g.fetchAssetsWithContext(ctx)` instead of `g.fetchAssets(ctx, logger, docChannel)` to make the code more readable and to clearly show that the function is using the context.
2. Use `select` statement to handle context cancellation and waiting for the interval, instead of using `time.Sleep` and `ctx.Err()` check.

``` go
func (g *githubCollector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
	if g.poll {
		for {
			select {
			case <-ctx.Done():
				return nil
			case <-time.After(g.interval):
				if g.tag != "" {
					return fmt.Errorf("Release tag should not be specified when using polling")
				}
				if err := g.fetchAssetsWithContext(ctx); err != nil {
					return err
				}
			}
		}
	} else {
		if g.tag == "" {
			return fmt.Errorf("Expected to receive release tag")
		}
		if err := g.fetchAssetsWithContext(ctx); err != nil {
			return err
		}
	}
	return nil
}

func (g *githubCollector) fetchAssetsWithContext(ctx context.Context) error {
	logger := logging.FromContext(ctx)
	err := g.fetchAssets(ctx, logger, docChannel)
	if err != nil {
		logger.Error("Error fetching assets", zap.Error(err))
		return fmt.Errorf("Error fetching assets: %v", err)
	}
	return nil
}
```